### PR TITLE
Document good pagination practices

### LINKED ISSUE #3582: Document good pagination practices
* Warn against offset pagination, explain keyset navigation (https://use-the-index-luke.com/no-offset, https://www.slideshare.net/MarkusWinand/p2d2-pagination-done-the-postgresql-way).
* Based on the outcome of https://github.com/dotnet/efcore/issues/26822, show patterns for using row value comparisons.

This can be a new page under the performance section, linked to from relevant query docs pages.

### PR DESCRIPTION
**Pull request by roji**
Note: this is a section under "efficient querying" under performance. But it could also be its own page under the top-level "Query data", since this isn't only about performance, but also about general correctness.

Closes #3582

### CODE REVIEW

File: entity-framework/core/performance/efficient-querying.md [hunk 1]
```diff
@@ -183,6 +183,43 @@ Here are the results for a benchmark comparing tracking vs. no-tracking behavior
...
+[!code-csharp[Main](../../../samples/core/Performance/Program.cs#KeySetPagination)]
+
+Assuming an index is defined on `PostId`, this query is very efficient, and also isn't sensitive to any concurrent changes happening in lower Id values.
+
+Keyset pagination is appropriate for pagination interfaces where the user navigates forwards and backwards, but does not support random access, where the user can jump to any specific page. Random access pagination typically requires using offset pagination as explained above.
+
+To learn more about the shortcomings of offset-based pagination and about keyset pagination, [see this post](https://use-the-index-luke.com/no-offset).
+
+> [!WARNING]
+> Always make sure that your ordering is fully deterministic. For example, if results are ordered only by date, but there can be multiple results with the same date, then results could be skipped when paginating as they're ordered differently across two queries. Ordering by both date and ID (or any other unique property) makes the resultset deterministic and avoids this problem.
```
Comment (ajcvickers) [inline entity-framework/core/performance/efficient-querying.md:210]: Possibly explicitly call out (again) that relational queries are not ordered by default.

File: entity-framework/core/performance/efficient-querying.md [hunk 2]
```diff
@@ -183,6 +183,43 @@ Here are the results for a benchmark comparing tracking vs. no-tracking behavior
...
+2. If any updates occur concurrently, your pagination may end up skipping certain entries or showing them twice. For example, if an entry is removed as the user is moving from page 2 to 3, the whole resultset "shifts up", and one entry would be skipped.
+
+The recommended alternative to offset-based pagination - sometimes called *keyset pagination* - is simply to remember the details of the last entry fetched (e.g. its ID, assuming you're ordering by that), and to ask for the next rows after that row. For example, assuming the last entry in the last page we fetched had an ID value of 55, we'd simply do the following:
+
+[!code-csharp[Main](../../../samples/core/Performance/Program.cs#KeySetPagination)]
+
+Assuming an index is defined on `PostId`, this query is very efficient, and also isn't sensitive to any concurrent changes happening in lower Id values.
+
+Keyset pagination is appropriate for pagination interfaces where the user navigates forwards and backwards, but does not support random access, where the user can jump to any specific page. Random access pagination typically requires using offset pagination as explained above.
```
Comment (ajcvickers) [inline entity-framework/core/performance/efficient-querying.md:205]: Is it worth suggesting a combination of both? For example, use an offset-based pagination for forward one or maybe a few pages, with skip/take to reset to a randomly jumped to page?

**Review by ajcvickers — APPROVED** [review]

File: entity-framework/core/performance/efficient-querying.md (hunk 2 above)
Comment (roji) [reply entity-framework/core/performance/efficient-querying.md:205]: Good idea, will do.

**Comment by roji** [comment]
@ajcvickers I moved this into its own page under the query section and did various improvements - can you take another look?

**Review by ajcvickers — APPROVED** [review]

**Comment by ajcvickers** [comment]
:shipit:

**Pull request merged**